packages/common/test/util.test.js

```diff
@@ -1,5 +1,365 @@
 import { expect } from 'chai';
-import { expandReferences, normalizeOauthConfig } from '../src/util';
+import {
+  enableMockClient,
+  expandReferences,
+  normalizeOauthConfig,
+  get,
+  post,
+  del,
+} from '../src/util';
+
+const client = enableMockClient('https://www.example.com');
+describe('request', () => {
+  it('should not set header content-type to application/json if body is string', async () => {
+    let request;
+    client
+      .intercept({
+        path: '/api',
+        method: 'POST',
+      })
+      .reply(200, r => {
+        request = r;
+        return '';
+      });
+
+    await post('/api', 'some string', {
+      baseUrl: 'https://www.example.com',
+    });
+
+    expect(request.headers).to.eql({});
+  });
+  it('should use baseUrl from options', async () => {
+    let request;
+    client
+      .intercept({
+        path: '/api',
+        method: 'GET',
+      })
+      .reply(200, r => {
+        request = r;
+        return {};
+      });
+
+    const response = await get('/api', { baseUrl: 'https://www.example.com' });
+
+    expect(request.path).to.eql('/api');
+    expect(response.code).to.eql(200);
+  });
+
+  it('should use fullUrlOrPath', async () => {
+    let request;
+    client
+      .intercept({
+        path: '/api',
+        method: 'GET',
+      })
+      .reply(200, r => {
+        request = r;
+        return {};
+      });
+
+    const response = await get('https://www.example.com/api');
+
+    expect(request.path).to.eql('/api');
+    expect(response.code).to.eql(200);
+  });
+});
+describe('methods', () => {
+  it('should send a GET', async () => {
+    // set up mock GET at /
+    let request;
+    client
+      .intercept({
+        path: '/api',
+        method: 'GET',
+      })
+      .reply(200, r => {
+        request = r;
+        return {};
+      });
+
+    await get('https://www.example.com/api');
+
+    // test the result
+    expect(request.method).to.eql('GET');
+  });
+
+  it('should send a POST', async () => {
+    // set up mock POST at /
+    let request;
+
+    client
+      .intercept({
+        path: '/api',
+        method: 'POST',
+      })
+      .reply(200, r => {
+        request = r;
+        return {};
+      });
+
+    await post('https://www.example.com/api');
+
+    // test the result
+    expect(request.method).to.eql('POST');
+  });
+
+  it('should send a DELETE', async () => {
+    // set up mock DELETE at /
+    let request;
+
+    client
+      .intercept({
+        path: '/api',
+        method: 'DELETE',
+      })
+      .reply(200, r => {
+        request = r;
+        return {};
+      });
+
+    await del('https://www.example.com/api');
+
+    // test the result
+    expect(request.method).to.eql('DELETE');
+  });
+});
+
+describe('options', () => {
+  it('should not throw 204 if response is okay using errorMap', async () => {
```

Comment:
Inside a describe block, you should have a series of tests which tell a story. Starting from the simplest, most general case, each test should build on the last one by adding complexity, or changing an edge-case, or altering the input. Ideally each test should just change a single thing about the input,

Also, because all (almost all?) these tests are about the error map, the describe block should be called `errorMap`.

The tests should be something like this:
```
should not throw for a 200+ error
should not throw for a 300+ error
should throw for a 400+ error
should throw for a 400+ error
should throw for a 200 error if a mapping is provided
should throw a custom error for a 404
should use a function to generate an error message
should ignore mappings that don't apply
```

The nice thing about these tests is that you can see the progression in complexity and it explains the behaviour of the error map, In fact, the tests start to look like documentation!